### PR TITLE
UIOR-1569: Correct refresh of prefix and suffix fields

### PR DESCRIPTION
**Pull request by charismatic-claire**
## Purpose

https://folio-org.atlassian.net/browse/UIOR-1569

This is a bugfix for the following problem: Prefix and suffix do not correctly refresh on selection in the create and edit of a Purchase Order.

## Approach

How did we solve this? Well, `FieldPrefix` and `FieldSuffix` contained broken validation logic, s.t. we could run into cases where we need to validate fields that aren't present. To fix that, we put the condition `shouldRevalidatePoNumber` in place, taking care of that cases. That makes sure we only run the validation when `poNumber` was mounted and is actually present, s.t. it can really be used for validation.

### CODE REVIEW

File: src/common/POFields/FieldPrefix.js [hunk 1]
```diff
@@ -13,13 +14,21 @@ const FieldPrefix = ({
   prefixes,
   ...rest
 }) => {
+  const { getFieldState } = useForm();
+  const poNumberFieldState = getFieldState?.(PO_FORM_FIELDS.poNumber);
+  const shouldRevalidatePoNumber = Boolean(
+    poNumberFieldState?.value !== undefined &&
+    poNumberFieldState?.value !== '' &&
+    poNumberFieldState?.value !== null,
```
Comment (elsenhans) [inline src/common/POFields/FieldPrefix.js]: can be simplified to `const shouldRevalidatePoNumber = poNumberFieldState?.value;`

**Comment by elsenhans** [comment]
Please add ticket to `CHANGELOG.md`

File: src/common/POFields/FieldPrefix.js (hunk 1 above)
Comment (charismatic-claire) [reply src/common/POFields/FieldPrefix.js]: Good point, simplified.

File: src/common/POFields/FieldPrefix.test.js [hunk 2]
```diff
@@ -1,9 +1,33 @@
...
+jest.mock('@folio/stripes-acq-components', () => {
+  const React = jest.requireActual('react');
+  const PropTypes = jest.requireActual('prop-types');
+
+  return {
+    FieldSelectFinal: ({ label, validateFields }) => (
+      <div data-testid="field-select" data-validate-fields={JSON.stringify(validateFields ?? [])}>
+        {label}
+      </div>
+    ),
```
Comment (elsenhans) [inline src/common/POFields/FieldPrefix.test.js]:
```suggestion
    FieldSelectFinal: (props) => mockFieldSelectFinal(props),
```

File: src/common/POFields/FieldPrefix.test.js [hunk 3]
```diff
@@ -1,9 +1,33 @@
...
 
+jest.mock('react-final-form', () => {
+  const actual = jest.requireActual('react-final-form');
+
+  return {
+    ...actual,
+    useForm: jest.fn(),
+  };
+});
+
```
Comment (elsenhans) [inline src/common/POFields/FieldPrefix.test.js]:
```suggestion
const mockFieldSelectFinal = jest.fn(({ label }) => label);

```

File: src/common/POFields/FieldPrefix.test.js [hunk 4]
```diff
@@ -21,9 +45,33 @@ const renderFieldPrefix = (props = {}) => render(
...
 
     expect(screen.getByText('ui-orders.orderDetails.orderNumberPrefix')).toBeInTheDocument();
   });
+
+  it('should not revalidate the PO number field when it is not mounted or empty', () => {
+    useForm.mockReturnValue({ getFieldState: () => undefined });
+
+    renderFieldPrefix();
+
+    expect(screen.getByTestId('field-select').dataset.validateFields).toBe('[]');
```
Comment (elsenhans) [inline src/common/POFields/FieldPrefix.test.js]: Please avoid the use of `data-testid` and `getByTestId`. Use DOM-based queries (e.g. `getByRole`, `getByLabelText`, `getByText`) instead or assert on the props passed to the mocked component.

```suggestion
expect(mockFieldSelectFinal).toHaveBeenCalledWith(expect.objectContaining({ validateFields: [] }));
```

File: src/common/POFields/FieldPrefix.test.js [hunk 5]
```diff
@@ -21,9 +45,33 @@ const renderFieldPrefix = (props = {}) => render(
...
+  });
+
+  it('should revalidate the PO number field when it is mounted and has a value', () => {
+    useForm.mockReturnValue({
+      getFieldState: () => ({ name: PO_FORM_FIELDS.poNumber, value: '1001' }),
+    });
+
+    renderFieldPrefix();
+
+    expect(screen.getByTestId('field-select').dataset.validateFields).toBe(JSON.stringify([PO_FORM_FIELDS.poNumber]));
```
Comment (elsenhans) [inline src/common/POFields/FieldPrefix.test.js]:
```suggestion
    expect(mockFieldSelectFinal).toHaveBeenCalledWith(expect.objectContaining({
      validateFields: [PO_FORM_FIELDS.poNumber],
    }));
```

File: src/common/POFields/FieldSuffix.test.js [hunk 6]
```diff
@@ -1,9 +1,33 @@
...
+  };
+});
+
+jest.mock('@folio/stripes-acq-components', () => {
+  const React = jest.requireActual('react');
+  const PropTypes = jest.requireActual('prop-types');
+
+  return {
+    FieldSelectFinal: ({ label, validateFields }) => (
+      <div data-testid="field-select" data-validate-fields={JSON.stringify(validateFields ?? [])}>
```
Comment (elsenhans) [inline src/common/POFields/FieldSuffix.test.js]: Same like for FieldPrefix.test.js:
> Please avoid the use of `data-testid` and `getByTestId`. Use DOM-based queries (e.g. `getByRole`, `getByLabelText`, `getByText`) instead or assert on the props passed to the mocked component.

File: src/common/POFields/FieldPrefix.test.js [hunk 7]
```diff
@@ -21,9 +45,33 @@ const renderFieldPrefix = (props = {}) => render(
...
+
+  it('should render the prefix field', () => {
+    useForm.mockReturnValue({ getFieldState: () => undefined });
+
     renderFieldPrefix();
 
     expect(screen.getByText('ui-orders.orderDetails.orderNumberPrefix')).toBeInTheDocument();
   });
+
+  it('should not revalidate the PO number field when it is not mounted or empty', () => {
```
Comment (elsenhans) [inline src/common/POFields/FieldPrefix.test.js]: Empty string is never actually tested. The test is named "_should not revalidate ... when it is not mounted or empty_", but it only covers the "not mounted" case, never the case where the field is mounted with value: ''. 
Please consider adding a test for an empty value and/or renaming the test.

File: src/common/POFields/FieldSuffix.test.js [hunk 8]
```diff
@@ -21,9 +45,33 @@ const renderFieldSuffix = (props = {}) => render(
...
+
+  it('should render the suffix field', () => {
+    useForm.mockReturnValue({ getFieldState: () => undefined });
+
     renderFieldSuffix();
 
     expect(screen.getByText('ui-orders.orderDetails.orderNumberSuffix')).toBeInTheDocument();
   });
+
+  it('should not revalidate the PO number field when it is not mounted or empty', () => {
```
Comment (elsenhans) [inline src/common/POFields/FieldSuffix.test.js]: Same here:
> Empty string is never actually tested. The test is named "_should not revalidate ... when it is not mounted or empty_", but it only covers the "not mounted" case, never the case where the field is mounted with value: ''.
Please consider adding a test for an empty value and/or renaming the test.

**Review by elsenhans — CHANGES_REQUESTED** [review]

**Comment by charismatic-claire** [comment]
@elsenhans Do you like the changes?

**Comment by elsenhans** [comment]
> @elsenhans Do you like the changes?

Looking good. Thanks for the additional test case.
Some small points:
1. Seems like line 20 is unused:
`const React = jest.requireActual('react');`
Please remove it.

2. Fix eslint errors in `FieldPrefix.test.js` and `FieldSuffix.test.js`

**Review by NikitaSedyx — CHANGES_REQUESTED** [review]
please address linter issues

additionally I didn't get how validation exclusion when po number is present solves your issue, I believe you just need to make component rerender and do not touch validation

**Comment by charismatic-claire** [comment]
@elsenhans Thanks for finding this, I'll address the issue.
@NikitaSedyx I will address the linter issues. On your second question: I think the thrown exceptions due to the throwing validation (because it cannot grab the value to validate) IS THE CAUSE of the issue of the field not refreshing properly. See the video attached. Otherwise I wouldn't expect that an input field update requires a manual refresh. The reactive nature of, well... React, should take care of that. I think.

[Screencast_20260723_123408.webm](https://github.com/user-attachments/assets/c3073a52-757e-4b0b-84cd-57f6d51281e9)

**Comment by charismatic-claire** [comment]
@elsenhans I adressed your comments.

@NikitaSedyx I fixed all the linter issues including two unrelated ones. I added an extra commit for it to indicate the fix doesn't belong to the ticket itself.

File: src/components/POLine/POLineAgreementLines/POLineAgreementLinesContainer.js [hunk 9]
```diff
@@ -49,7 +49,7 @@ const POLineAgreementLinesContainer = ({ lineId, label, mutator }) => {
     [fetchAgeementLines],
   );
 
-  useEffect(refreshList, [lineId]);
+  useEffect(refreshList, [lineId, refreshList]);
```
Comment (charismatic-claire) [inline src/components/POLine/POLineAgreementLines/POLineAgreementLinesContainer.js]: @NikitaSedyx out of scope linter fix in extra commit. If that's considered bad practice, we can remove it.

File: src/components/PurchaseOrder/POForm.js [hunk 10]
```diff
@@ -120,7 +120,7 @@ const POForm = ({
         ...(orderTemplate?.hiddenFields || {}),
       }));
     }
-  }, []);
+  }, [initialValues.template, parentResources?.orderTemplates?.records]);
```
Comment (charismatic-claire) [inline src/components/PurchaseOrder/POForm.js]: @NikitaSedyx out of scope linter fix in extra commit. If that's considered bad practice, we can remove it.

Comment (NikitaSedyx) [reply src/components/PurchaseOrder/POForm.js]: I would suggest to not fix issues not connected to your changes

**Comment by NikitaSedyx** [comment]
> I think the thrown exceptions due to the throwing validation (because it cannot grab the value to validate) IS THE CAUSE of the issue of the field not refreshing properly. See the video attached. Otherwise I wouldn't expect that an input field update requires a manual refresh. The reactive nature of, well... React, should take care of that. I think.

hm, in some extent it makes sense, but I would like to play a bit - it's pretty old code so I do not remember all details

**Comment by charismatic-claire** [comment]
@NikitaSedyx did you take a look and could reproduce the bug and verify the fix?

File: src/components/PurchaseOrder/POForm.js (hunk 10 above)
Comment (charismatic-claire) [reply src/components/PurchaseOrder/POForm.js]: Alrighty

**Review by NikitaSedyx — APPROVED** [review]

**Review by elsenhans — APPROVED** [review]
LG
